app/controllers/concerns/api/import_puppetclasses_common_controller.rb

```diff
@@ -39,10 +40,16 @@ def import_puppetclasses
     end
 
     # RUN PuppetClassImporter
-    if (errors = ::PuppetClassImporter.new.obsolete_and_new(@changed)).empty?
-      render("api/v1/import_puppetclasses/#{rabl_template}", :layout => "api/layouts/import_puppetclasses_layout")
+    if params.key?(:background) && !['false', false].include?(params[:background])
+      task = ForemanTasks.async_task(::Actions::Foreman::PuppetClass::Import, :changed => @changed)
+      process_success task
     else
-      render :json => {:message => _("Failed to update the environments and Puppet classes from the on-disk puppet installation: %s") % errors.join(", ")}, :status => :internal_server_error
+      begin
+        task = ForemanTasks.sync_task(::Actions::Foreman::PuppetClass::Import, :changed => @changed)
```

Comment:
You can use trigger_task https://github.com/theforeman/foreman-tasks/blob/master/lib/foreman_tasks.rb#L21, that takes the `async` as parameter and we would not need the `if/else` condition here